feat: Add labels to node information in the application pod view (#15274)

msoderberg opened the pull request:
The goal of this PR is to add support for optionally showing node labels in the application pod view. 

Closes #15274 

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

blakepettersson commented on controller/appcontroller.go:
```diff
@@ -617,7 +617,7 @@ func (ctrl *ApplicationController) getAppHosts(a *appv1.Application, appNodes []
 			return resourcesInfo[i].ResourceName < resourcesInfo[j].ResourceName
 		})
-		hosts = append(hosts, appv1.HostInfo{Name: nodeName, SystemInfo: node.SystemInfo, ResourcesInfo: resourcesInfo})
+		hosts = append(hosts, appv1.HostInfo{Name: nodeName, SystemInfo: node.SystemInfo, ResourcesInfo: resourcesInfo, Labels: node.Labels}) //TODO: Only append labels defined in conf
```
My spontaneous thought is that this could be managed on an `AppProject` level, i.e. to have a whitelist of node labels as an attribute of an `AppProject`.

Something like

```yaml
allowedNodeLabels:
   - kubernetes.io/arch
   - kubernetes.io/os
   - topology.kubernetes.io/zone
   - node.kubernetes.io/instance-type
   - karpenter.k8s.aws/instance-cpu
   - karpenter.k8s.aws/instance-memory
```

WDYT @crenshaw-dev?

blakepettersson commented:
@msoderberg I see that you implemented my idea 😄 

Does anyone have any objections to this approach (e.g @rbreeze, @crenshaw-dev)?

msoderberg commented:
> @msoderberg I see that you implemented my idea 😄 
> 
> 
> 
> Does anyone have any objections to this approach (e.g @rbreeze, @crenshaw-dev)?

Yes 🙂 I think it makes sense 👍 

The backend changes are ready for review now, but there's still work to be done on the ui.

msoderberg commented:
I've updated the UI now and the PR is ready for review. Here are some screenshots:

Project settings:
![project-settings](https://github.com/argoproj/argo-cd/assets/5403544/fe64e100-d9d5-4df1-98d9-f60833038f35)
Pod view:
![pod-view](https://github.com/argoproj/argo-cd/assets/5403544/ac4addd8-2ec7-4081-be5b-7e4fb9519d61) 
Multiple labels (max height and scrollable):
![pod-view-max-height](https://github.com/argoproj/argo-cd/assets/5403544/16e39ab5-d702-47ca-9df5-dd820d561cef) 

Below is an example where the label only exist on one of the nodes. It also shows an issue with long k/v pairs that I haven't solved yet. I'm a bit lost when it comes to CSS, so I'm open for suggestions on how to make it look better :) 
![pod-view-long-key-and-value](https://github.com/argoproj/argo-cd/assets/5403544/65e77cb3-98ec-48f5-b773-0921336f9e8d)

blakepettersson commented on cmd/argocd/commands/project.go:
```diff
@@ -78,6 +79,8 @@ func NewProjectCommand(clientOpts *argocdclient.ClientOptions) *cobra.Command {
 	command.AddCommand(NewProjectWindowsCommand(clientOpts))
 	command.AddCommand(NewProjectAddOrphanedIgnoreCommand(clientOpts))
 	command.AddCommand(NewProjectRemoveOrphanedIgnoreCommand(clientOpts))
+	command.AddCommand(NewProjectAdddNodeLabelCommand(clientOpts))
```
```suggestion
	command.AddCommand(NewProjectAddNodeLabelCommand(clientOpts))
```

blakepettersson commented on cmd/argocd/commands/project.go:
```diff
@@ -999,3 +1002,88 @@ func NewProjectEditCommand(clientOpts *argocdclient.ClientOptions) *cobra.Comman
 	return command
 }
+
+// NewProjectAddNodeLabelCommand returns a new instance of an `argocd proj add-node-label` command
+func NewProjectAdddNodeLabelCommand(clientOpts *argocdclient.ClientOptions) *cobra.Command {
```
```suggestion
// NewProjectAddNodeLabelCommand returns a new instance of an `argocd proj add-node-label` command
func NewProjectAddNodeLabelCommand(clientOpts *argocdclient.ClientOptions) *cobra.Command {
```

blakepettersson commented on docs/user-guide/projects.md:
```diff
@@ -7,6 +7,7 @@ teams. Projects provide the following features:
 * restrict where apps may be deployed to (destination clusters and namespaces)
 * restrict what kinds of objects may or may not be deployed (e.g. RBAC, CRDs, DaemonSets, NetworkPolicy etc...)
 * defining project roles to provide application RBAC (bound to OIDC groups and/or JWT tokens)
+* allow node labels to be propogated to the UI (they will be visible on nodes in application pod view)
```
```suggestion
* allow node labels to be propagated to the UI (they will be visible on nodes in application pod view)
```

blakepettersson commented on docs/user-guide/projects.md:
```diff
@@ -369,3 +370,20 @@ spec:
+
+## Allowed Node Labels
+
+Projects allow you to propogate node labels to the pod view of applications belonging to the project.
```
```suggestion
Projects allow you to propagate node labels to the pod view of applications belonging to the project.
```

blakepettersson commented:
I think this is great! Apart from a few minor nits, I have a few points 

* For really long labels, it might be worth adding a `text-overflow: ellipsis` rule, so that it doesn't break the tooltip design. 
* Alternatively, to aid discoverability it might be better to have it in the table as before, but with an overflow rule so that long labels do not bork the design.

Pending a sign-off from an approver that the general design is the way to go, I'd also add that for the CLI commands it would be good to have some unit tests for them (they haven't generally been tested but would be good to start). I think #12425 might have a good example of CLI tests

msoderberg commented:
> I think this is great! Apart from a few minor nits, I have a few points
> 
>     * For really long labels, it might be worth adding a `text-overflow: ellipsis` rule, so that it doesn't break the tooltip design.
> 
>     * Alternatively, to aid discoverability it might be better to have it in the table as before, but with an overflow rule so that long labels do not bork the design.
> 
> 
> Pending a sign-off from an approver that the general design is the way to go, I'd also add that for the CLI commands it would be good to have some unit tests for them (they haven't generally been tested but would be good to start). I think #12425 might have a good example of CLI tests

Thanks @blakepettersson! I've fixed the typos and added  a couple of initial tests for the project commands (similar to how it's done for the applicationset commands). I plan to add tests for the actual add/remove label logic as well, but I'll probably pause my work on this until we have an OK on the design.

blakepettersson commented:
@pasha-codefresh Git conflicts notwithstanding, does this make sense to you?

msoderberg commented:
I decided to revisit this old PR as it's still something that would be useful for us. I've sorted out the conflicts and decided to revert back to showing the labels in the node info table instead of in tooltips. Made a couple of small tweaks to handle long label keys and values as well. I think it's usable now, but it can most definitely be improved :)

agaudreault commented on cmd/argocd/commands/project.go:
```diff
@@ -84,6 +85,8 @@ func NewProjectCommand(clientOpts *argocdclient.ClientOptions) *cobra.Command {
 	command.AddCommand(NewProjectRemoveSourceNamespace(clientOpts))
 	command.AddCommand(NewProjectAddDestinationServiceAccountCommand(clientOpts))
 	command.AddCommand(NewProjectRemoveDestinationServiceAccountCommand(clientOpts))
+	command.AddCommand(NewProjectAddNodeLabelCommand(clientOpts))
```
I dont think the configuration should be on a per-project basis. The labels are usually different per clusters, but this sounds more like an admin configuration rather than a project restriction.

agaudreault reviewed (changes requested):
Rejecting just because I think this needs more discussions on how to implement the feature in the issue before merging the code. Lets move the discussion back to the issue and proposed ways to configure this feature before implementing changes.

(the code quality looks good however) :)
